### PR TITLE
fix(write_source_files): fix writing to workspace root

### PR DESCRIPTION
**Pull request by jbedard**

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -20,3 +20,16 @@ bzl_library(
...
+
+# write_source_files() to a git ignored subdirectory of the root
+load("//lib:write_source_files.bzl", "write_source_files")
+genrule(
+    name = "write_source_file_root",
+    outs = ["write_source_file-root_directory"],
+    cmd = "mkdir -p $@ && echo 'test' > $@/test.txt",
+    visibility = ["//visibility:private"],
+)
+write_source_files(
```
Comment (jbedard) [inline BUILD.bazel:32]: @gregmagolan did you actually get your `write_dist` running as a test? This only failed previously when doing `bazel run` and not `bazel test`...

File: BUILD.bazel [hunk 2]
```diff
@@ -20,3 +20,17 @@ bzl_library(
...
+    name = "write_source_file_root-test",
+    diff_test = False,
+    files = {"test-out/dist/write_source_file_root-test": ":write_source_file_root"},
+)
```
Comment (kormide) [inline BUILD.bazel:36]: Want to add `//visibility:private` to this as well?

**Comment by kormide** [comment]
Nothing actually runs this test, right? I'm wondering if we should just leave it out of WORKSPACE.

**Comment by kormide** [comment]
> Nothing actually runs this test, right? I'm wondering if we should just leave it out of WORKSPACE.

Or at least call it out as a manual test that ought to be some sort of e2e test.

**Comment by jbedard** [comment]
This is the same as what @gregmagolan did in `lib/tests/write_source_files:write_dist`. I think we should find a way to actually run them in CI...

File: BUILD.bazel (hunk 2 above)
Comment (jbedard) [reply BUILD.bazel:36]: Shouldn't tests be public?

Comment (kormide) [reply BUILD.bazel:36]: Hm, the angle I was gettting at was whether this should be available to anyone who depends on our repository.

**Comment by kormide** [comment]
> This is the same as what @gregmagolan did in `lib/tests/write_source_files:write_dist`. I think we should find a way to actually run them in CI...

That should be pretty easy to do. Just add another run step in `.github/workflows/ci.yaml`.
```yaml
- name: manual integration test
   run: bazel run //:write_source_file_root-test
```

File: lib/private/write_source_file.bzl [hunk 3]
```diff
@@ -266,7 +266,7 @@ def _write_source_file_impl(ctx):
         else:
             fail("in file %s must be a single file or a target that provides DefaultOutputPathInfo or DirectoryPathInfo" % ctx.attr.in_file.label)
 
-        out_path = "/".join([ctx.label.package, ctx.attr.out_file])
+        out_path = "/".join([ctx.label.package, ctx.attr.out_file]) if len(ctx.label.package) > 0 else ctx.attr.out_file
```
Comment (gregmagolan) [inline lib/private/write_source_file.bzl:269]: I believe you can shorten this to
```
out_path = "/".join([ctx.label.package, ctx.attr.out_file]) if ctx.label.package else ctx.attr.out_file
```
empty string is a falsy in starlark

File: BUILD.bazel (hunk 2 above)
Comment (gregmagolan) [reply BUILD.bazel:36]: It is private by default unless there is a default_visiblity in the BUILD to set another default

File: BUILD.bazel (hunk 1 above)
Comment (gregmagolan) [reply BUILD.bazel:32]: Needs an integration test but I'd prefer to avoid the added complexity of bazel-in-bazel

**Review by kormide — APPROVED** [review]
